## Objective
Explore the cleaned Austin Animal Center dog dataset to identify patterns,
relationships, and feature distributions relevant to predicting shelter stay duration.

**Data source:** `shelter_data.pkl` (produced in `01_data_preparation.ipynb`)

In [1]:
# Standard library
import datetime

# Third-party
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# Configuration
sns.set_theme(style="whitegrid")
plt.rcParams["figure.dpi"] = 100
pd.set_option("display.max_columns", None)
pd.set_option("display.float_format", "{:,.2f}".format)

In [2]:
df = pd.read_pickle("data/shelter_data.pkl")
print(f"Shape: {df.shape[0]:,} rows × {df.shape[1]} columns")
print(f"Date range: {df['DateTime_in'].min():%Y-%m-%d} → {df['DateTime_in'].max():%Y-%m-%d}")
df.head()

Shape: 87,176 rows × 25 columns
Date range: 2013-10-01 → 2023-12-03


,Animal ID,Name,DateTime_in,Found Location,Intake Type,Intake Condition,Breed,Color,Month_in,Year_in,Age upon Intake (months),Number of Intake,Left Shelter,Status_in,Sex_in,Month_out,Year_out,Age upon Outcome (months),DateTime_out,Number of Outcome,Outcome Type,Outcome Subtype,Status_out,Sex_out,Time in shelter
0,A006100,Scamp,2014-03-07 14:26:00,8700 Research in Austin (TX),Public Assist,Normal,Spinone Italiano Mix,Yellow/White,3,2014,72,1,True,Neutered,Male,3.0,2014.0,72,2014-03-08 17:10:00,1,Return to Owner,NaN,Neutered,Male,1 days 02:44:00
1,A006100,Scamp,2014-12-19 10:21:00,8700 Research Blvd in Austin (TX),Public Assist,Normal,Spinone Italiano Mix,Yellow/White,12,2014,84,2,True,Neutered,Male,12.0,2014.0,84,2014-12-20 16:35:00,2,Return to Owner,NaN,Neutered,Male,1 days 06:14:00
2,A006100,Scamp,2017-12-07 14:07:00,Colony Creek And Hunters Trace in Austin (TX),Stray,Normal,Spinone Italiano Mix,Yellow/White,12,2017,120,3,True,Neutered,Male,12.0,2017.0,120,2017-12-07 00:00:00,3,Return to Owner,NaN,Neutered,Male,0 days 05:00:00
3,A047759,Oreo,2014-04-02 15:55:00,Austin (TX),Owner Surrender,Normal,Dachshund,Tricolor,4,2014,120,1,True,Neutered,Male,4.0,2014.0,120,2014-04-07 15:12:00,1,Transfer,Partner,Neutered,Male,4 days 23:17:00
4,A134067,Bandit,2013-11-16 09:02:00,12034 Research Blvd in Austin (TX),Public Assist,Injured,Shetland Sheepdog,Brown/White,11,2013,192,1,True,Neutered,Male,11.0,2013.0,192,2013-11-16 11:54:00,1,Return to Owner,NaN,Neutered,Male,0 days 02:52:00


### Data Integrity Validation
Verify that the pickled DataFrame retains all expected dtypes and no rows
were lost during serialization.

In [3]:
# Quick dtype + missing value overview
dtype_summary = pd.DataFrame({
    "dtype": df.dtypes,
    "missing_n": df.isna().sum(),
    "missing_%": (df.isna().sum() / len(df) * 100).round(2),
    "unique_n": df.nunique()
})
dtype_summary

,dtype,missing_n,missing_%,unique_n
Animal ID,str,0,0.00,73783
Name,str,13765,15.79,17980
DateTime_in,datetime64[us],0,0.00,68727
Found Location,str,0,0.00,41624
Intake Type,str,0,0.00,6
Intake Condition,str,0,0.00,18
Breed,str,0,0.00,2554
Color,str,0,0.00,392
Month_in,int32,0,0.00,12
Year_in,int32,0,0.00,11


### Missing Data Assessment

The most significant missing values appear in:

**Name (15.8%)**: Animals often enter the shelter unnamed; treated as acceptable

**Outcome Subtype (67.2%)**: High missing rate indicates many outcomes lack granular classification. This limits detailed subtype analysis but doesn't prevent primary Outcome Type analysis.

**Time in shelter (7.8%)**: Target variable with missing values - 

Missing values are preserved for now and addressed during specific analyses.